tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Session.java

```diff
@@ -432,6 +434,106 @@ public Runner runner() {
+   * operation, which is {@link Init#DEFAULT_NAME}. For example:</p>
+   * <pre>{@code
+   * try (Graph g = new Graph()) {
+   *   Ops tf = Ops.create(g);
+   *
+   *   Variable<TInt32> x = tf.variable(tf.constant(10));
+   *   Variable<TInt32> y = tf.variable(tf.constant(20));
+   *   Add<TInt32> z = tf.math.add(x, y);
+   *   tf.init();
+   *
+   *   try (Session s = new Session(g)) {
+   *     s.runInit();
+   *
+   *     try (Tensor<TInt32> t = s.runner().fetch(z).run().get(0).expect(TInt32.DTYPE)) {
+   *       assertEquals(30, t.data().getInt());
+   *     }
+   *   }
+   * }
+   * }</pre>
+   */
+  public void runInit() {
+    runInit(Init.DEFAULT_NAME);
+  }
+
+  /**
+   * Run all graph initializers grouped under the {@code initOpName} operation.
+   *
+   * <p>Initializers must be executed once before running the graph in a training loop using a
+   * session {@link Runner}.</p>
+   *
+   * <p>The {@code initOpName} is the name of a single operation already added to the
+   * graph that executes all graph initializers at once. For example:</p>
+   * <pre>{@code
+   * try (Graph g = new Graph()) {
+   *   Ops tf = Ops.create(g);
+   *
+   *   Variable<TInt32> x = tf.variable(tf.constant(10));
+   *   Variable<TInt32> y = tf.variable(tf.constant(20));
+   *   Add<TInt32> z = tf.math.add(x, y);
+   *   tf.withName("initialize").init();
+   *
+   *   try (Session s = new Session(g)) {
+   *     s.runInit("initialize");
+   *
+   *     try (Tensor<TInt32> t = s.runner().fetch(z).run().get(0).expect(TInt32.DTYPE)) {
+   *       assertEquals(30, t.data().getInt());
+   *     }
+   *   }
+   * }
+   * }</pre>
+   *
+   * @param initOpName name of the initializer operation.
+   */
+  public void runInit(String initOpName) {
+    Operation operation = graph.operation(initOpName);
+    if (operation == null) {
+      throw new IllegalArgumentException("Initializer operation named '"
+          + initOpName + "' cannot be found in the graph");
+    }
+    runner().addTarget(operation).run();
+  }
+
+  /**
+   * Run all graph initializers registered by the given {@code init} op.
+   *
+   * <p>Initializers must be executed once before running the graph in a training
+   * loop using a session {@link Runner}.</p>
+   *
+   * <p>This method can be used if the graph is being built by the same program running the session.
+   * For example:</p>
+   * <pre>{@code
+   * try (Graph g = new Graph()) {
+   *   Ops tf = Ops.create(g);
+   *
+   *   Variable<TInt32> x = tf.variable(tf.constant(10));
+   *   Variable<TInt32> y = tf.variable(tf.constant(20));
+   *   Add<TInt32> z = tf.math.add(x, y);
+   *   Init init = tf.withName("initialize").init();
+   *
+   *   try (Session s = new Session(g)) {
+   *     s.runInit(init);
+   *
+   *     try (Tensor<TInt32> t = s.runner().fetch(z).run().get(0).expect(TInt32.DTYPE)) {
+   *       assertEquals(30, t.data().getInt());
+   *     }
+   *   }
+   * }
+   * }</pre>
+   */
+  public void runInit(Init initOp) {
```

Comment:
Thanks for these changes Karl, I think these will be super helpful! Quick thought:

Since the new `Init` class is a `RawOp`, do we gain anything by adding the `runInit` method, as opposed to running the `RawOp` in the normal fashion:

```
session.runner().addTarget(initOp).run()
```
We could even add additional `run` methods to Session that accept single `Op`/`Operand`/etc. arguments, and have them automatically added as targets, to achieve syntax like 

```session.run(initOp)```


TensorFlow Python (pre 2.0) has a standard syntax for the variable initializers (for both [global](https://github.com/tensorflow/tensorflow/blob/fd88bbf95dc7e5b027dc0ff3eedbf3072b72e376/tensorflow/python/ops/variables.py#L3213) and [local](https://github.com/tensorflow/tensorflow/blob/fd88bbf95dc7e5b027dc0ff3eedbf3072b72e376/tensorflow/python/ops/variables.py#L3235) variables)

```
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
 
    ...
```
Maybe we could modify the Java API to match? It may not yet be on our roadmap to distinguish between `global` (shared across processes) variables and `local` (per-process) variables, but we could perhaps try for something like:

```
try (Graph graph = new Graph()) {
  Ops tf = Ops.create(graph);

  ...

  try (Session session = new Session(graph)) {
    session.addTarget(tf.variablesInitializer()).run(); // or session.run(tf.variablesInitializer)

    ...

  }
}
```

`tf.variablesInitializer()` can call `Graph.initializers()` that you added below. To make it even simpler, we could automatically add the `Assign` op to a graph when `tf.variable` is called with an initial value, so users don't need to keep track of adding each initialization to the initializers list.

The [global](https://github.com/tensorflow/tensorflow/blob/fd88bbf95dc7e5b027dc0ff3eedbf3072b72e376/tensorflow/python/ops/variables.py#L3039) vs. [local](https://github.com/tensorflow/tensorflow/blob/fd88bbf95dc7e5b027dc0ff3eedbf3072b72e376/tensorflow/python/ops/variables.py#L3090) variable distinction may be something we want to add later on though...

What do you think?

Reply:
Thanks @dhruvrajan ,

> We could even add additional run methods to Session that accept single Op/Operand/etc. arguments

Yes, I thought of having a `session.run(Op)` as well, I thought that by typing it we kind of "guide" the user to use it only for running initializers but I agree we can relax this a bit.

> tf.variablesInitializer() can call Graph.initializers() that you added below. 

The `tf.variableInitializers()` sounds pretty much like the `tf.init()`, no? But the problem that arise and that @Craigacp mentioned to me before is when you load a graph from disk, your graph won't have any initializers so you can't build that op at runtime, you need to find it in the graph.

That is why the `tf.init()` is always called before running the session. It's kind of annoying, I agree, to always remember calling this method before freezing the graph, I don't know how we can make it more obvious... I can also revert to what Adam did before, where we could retrieve the init op from the `Graph` instead of `Ops` (i.e. we would do something like `session.run(graph.variableInitializers())`). At least that simple `run` method would be a simple improvement.

>To make it even simpler, we could automatically add the Assign op to a graph when tf.variable

This is already supported! Check [here](https://github.com/tensorflow/java/blob/19c6dc8881e601dae9fe6a1ab4b24d0acef050d3/tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java#L7346)

Reply:
> This is already supported! Check here

Ah thanks for pointing that out, awesome!

>  when you load a graph from disk, your graph won't have any initializers so you can't build that op at runtime, you need to find it in the graph.

Hmm this is interesting! If you load a graph from disk, why won't the graph have any initializers? (It may have init ops in the graph, I guess, but they wouldn't be mapped to the `initializers` list we maintain in Java?). In this case, could we just create an `Init` Op that does nothing?

> That is why the tf.init() is always called before running the session.

To clarify, must the `Init` op be created before a session is created, or just run before the first `session.run`? If it is the latter, we can have:

```
try (Session session = new Session(graph)) {
  session.run(tf.init()); // maybe rename to variablesInitializer() for consistency?
}
```

Brainstorming other ways to keep the syntax consistent with Python, I had a couple ideas, eager to know your thoughts @karllessard:

1. If it really needs to go before `Session.create`, add the `Init` creation (`tf.init()`) call as the first step of `Session.create()` or a `tf.session()` factory method.

I can't think of a case where we would *not* want this Op to be added to the graph; this could also eliminate the need for the user to call `tf.init()`?

2. As in TensorFlow Python keep track of individual variable `initial_values` and `initializer` ops as fields within the `Variable` class. 

Thus instead of keeping track of a list of `initializer` ops on the graph, we only keep track of variables. We can extract the initializers by mapping over the variables. (See: [global_variables](https://github.com/tensorflow/tensorflow/blob/fd88bbf95dc7e5b027dc0ff3eedbf3072b72e376/tensorflow/python/ops/variables.py#L3039), [global_variables_initializer](https://github.com/tensorflow/tensorflow/blob/fd88bbf95dc7e5b027dc0ff3eedbf3072b72e376/tensorflow/python/ops/variables.py#L3039))

Then we get syntax and functionality super close to standard TF (I think, quite desirable)
```
try (Session session = tf.session()) { // automatically runs tf.init()
    session.run(tf.variablesInitializer());
}
```

Thanks for the detailed discussion! 😃

Reply:
The global_variables_initializer needs to be created in the same way that tf.init does at the moment.

Reply:
My concern of naming it `variableInitializers` instead of just `init` is that I think there might be other op that must be called prior to run a session that variable `assign`. 

For instance, in this [old example](https://github.com/karllessard/tensorflow-models/blob/20b13c2227b66cef4cb018220797e11df62a8422/samples/languages/java/suite/src/main/java/org/tensorflow/model/sample/mnist/MnistWithSummaries.java#L88) I had to run the `createSummaryFileWriter` as well. 

So maybe that name is not "right" in Python as well but they kept if for backward compatibility. Or the have more than one op ran before a session starts. 

Now for calling `tf.init()` implicitly at session creation or explicitly... I don't have the answer right now, to be honest, it is a tough call that requires more thinking. If @Craigacp says right that the actual behaviour mimics what is done in Python, then I would be comfortable to merge it as is and maybe find other ways to improve it later?

Reply:
Yup I agree; good idea to merge this, and update things reactively in the future. 

For now, then, the two main things I see that will be different between the current implementation and the Python implementation are:

1. In Java, we must call `tf.init()` before creating a session; in Python, the equivalent function can be called after a session is created.

2. In Java, we explicitly keep track of initializers within the `Graph` object. In Python, initializers are held by `Variable` objects, and retrieved when needed.

Let's just keep these differences in mind and see if anything related comes up in the future!

Reply:
Thinking of it, for point 1., I’m gonna test again, maybe it’s ok to add an op to a graph after a session is created, so you can simply call `session.run(tf.init())` if the graph is built in the same process as the session runner.

Reply:
Yep, looks like this works too!